### DISCUSSION TITLE
Variables implementation

### DISCUSSION THREAD

**Discussion by tomlynchRNA** [rnadigital/agentcloud#559] [General]
re #490

**Comment by tomlynchRNA** [comment]
Some important scenarios we need to consider for variables, all closely related:

- Dynamic session variables

Dynamic variables would be variables that can have their value set or updated during a session. For example, a process app may have a task description that uses a variable e.g. {question_answer} where that answer is taken from a previous tasks human input question, see next point.

- Human input variable output

Human input tasks should allow the selection of a named variable (or a variable per question for "form" human inputs) that the response to the human input will be saved into, and be accessible throughout the rest of the sessions prompts (tasks, agent role/goal, tool descriptions etc).

- Handling dynamic variables in agent-backend

Variables may need to change after a crew is "built" or after tools, etc are instantiated. i.e we can't just `whatever.prompt.format(**variables)` once, we need to do it when ever the variables are changed, or perhaps every time that prompt, task, tool, etc is used.

- (specifically for RAG tools) Ability to define "filters" in the tool parameters section. This is different from "parameters" (input to the function), and environment variables, but these are filters e.g. `WHERE column = '{our_variable}'`, so we need to define key, type of filter, value OR a json input box to input it raw, also must support variables. Variables would not be mandatory for this. We might break this into a separate ticket, but I mention it here because it's closely related.

-----

And as a side note, we should consider how to improve human input in chat apps to allow more formatted inputs and the ability to direct outputs for some responses to variables. This is difficult because of the lack of tasks in chat apps. @anada10

**Reply by tomlynchRNA** [reply]
Action items, feel free to come up with some of your own:
- [x] Check crewAI source code to see how it does variable resolution/interpolation into prompts and whether its realistic to update this to pull from a database, or in-memory state that lives for the length of the session each time a prompt is used. 
**good answer from past research by @iandjx  below in the replies :+1:** 
- [ ] think about where to store variables that are set during a session. perhaps a mapping of variable name or Id to the values on the session document. The current method of storing an array of variableIds, fetching them, then interpolating them may be problematic if you can have 2 of the same named variable. while it might not be a poblem to have two variables named the same thing, even if used within the same crew, for ones that we allow to be per-session, this would caues a conflict.
- [ ] Do the "dynamic" variables have to exist in the variables collection already, and do they need a variable ID? or can they simply be created by inputting a name in a human input task and using the same name later in the session.
- Think about how to communicate to the frontend that certain variables have a value or need to be given a value. this links closely to human input tasks -- try not to rely on the existing method of pulling the task by name (which would contain the human input form schema), but instead send messages from the agent-backend that signal the frontend to show the form based purely on the data in the message
- [ ] Think about how this links with out API. RE the points above, if the message requesting variable/human input is part of the chat message, that works better for API access than the current approach. Also, for example, if variables are on the session users can access the session via API to get the current session variables state.
- [ ] Does this step on the `context` concept, and can we leverage that along with structured output for some similar capabilities without such drastic changes? I think maybe, but dynamic variables make the order and structure of tasks less clunky because we don't have to interleave so many human input tasks and can use variables from previous tasks more selectively and arbitrarily.
- [x] Can we use the extra search_params/kwargs (forgot the exact name) in langchain vector stores to use these variables in filters in a simple way, or d owe have to construct the filter in a vector-db-specific way? **Answer: its vector db specific, but is actually pretty simple, see https://github.com/rnadigital/agentcloud/discussions/559#discussioncomment-10633054**

**Reply by iandjx** [reply]
I made some initial research on variable interpolation in crew-ai

`https://rna-digital.slack.com/archives/CQTU159M1/p1725505714606369?thread_ts=1725495769.730399&cid=CQTU159M1`

**Reply by tomlynchRNA** [reply]
Very nice, exactly what I wanted to know :pray: 
For posterity:

> looking at crew ai code
>     def interpolate_inputs(self, inputs: Dict[str, Any]) -> None:
>         """Interpolate inputs into the agent description and backstory."""
>         if self._original_role is None:
>             self._original_role = self.role
>         if self._original_goal is None:
>             self._original_goal = self.goal
>         if self._original_backstory is None:
>             self._original_backstory = self.backstory
> 
>         if inputs:
>             self.role = self._original_role.format(**inputs)
>             self.goal = self._original_goal.format(**inputs)
>             self.backstory = self._original_backstory.format(**inputs)
> interpolate_inputs
> function is called when variables are passed as arguments to the kickoff function.
> these would mean that placedholders e.g. {name}  will be replaced with the variable value or null
> we would have to store the original property value e.g. descriptoion, backstory, role, etc. such that everytime the variable value changes. we can use these as references for interpolation again. We have to do this because running interpolate_inputs
> deletes that placeholders that we use as reference for interpolation
> we would also need to expose interpolate_inputs so that these can be called in a task’s callback to pass in the new variable values

**Reply by tomlynchRNA** [reply]
Agreed that we would need to store the original prompt w/ placeholders still in there as a first step.

And we can think about where to call interpolate_inputs again later :+1:

**Reply by iandjx** [reply]
It would make sense to update the variables during the task callback. For us to easily determine which variable to update, we would have to use structured output feature of expected output so that we can easily parse what variables need to be updated

**Reply by tomlynchRNA** [reply]
We would already have the exact answer to the human input question(s) and variable associated with it from the form, so we can handle those however we want without necessarily requiring structured output, I think :slightly_smiling_face:

**Reply by iandjx** [reply]
Okay. But what about task output as arguments to change session variables?

**Reply by tomlynchRNA** [reply]
Ah, I see what you mean. Well I guess we could have both :sweat_smile: UX is another story, but both are possible :+1:

**Comment by iandjx** [comment]
What is the advantage of using dynamic session variables vs using human input/task output as context for the next tasks? Can we have some examples on when we would want to update the session variables mid-session?

**Reply by anada10** [reply]
Context is a crew/process app specific feature @iandjx.. we are designing a feature that will work across the platform across all parts of the app in a reliable and consistent manner, not just something for one subcomponent.

**Reply by tomlynchRNA** [reply]
another thing, right now if we ask a user to answer human input for e.g. "what is your name", in a later prompt, we rely 100% on LLM to know the users name and interpolate it into some answer, tool call parameter, etc later on. If this were a variable we have better control over it and could precisely take the a variable and use it as the input in later stages.

**Reply by anada10** [reply]
The advantage of a feature like this is quite evident when attempting to conduct a reliable RAG search.
For example, storing of variables would allow users to set form fields to be equal to metadata filters on a RAG search.
This is more reliable than using approaches like self query retrieval which tries to invent what value should be passed to the Vector store.

**Reply by iandjx** [reply]
> another thing, right now if we ask a user to answer human input for e.g. "what is your name", in a later prompt, we rely 100% on LLM to know the users name and interpolate it into some answer, tool call parameter, etc later on. If this were a variable we have better control over it and could precisely take the a variable and use it as the input in later stages.

this makes sense. @anada10 showed me issues before where the AI asked redundant questions even when context was provided earlier

**Comment by tomlynchRNA** [comment]
RE:
> Can we use the extra search_params/kwargs (forgot the exact name) in langchain vector stores to use these variables in filters in a simple way, or d owe have to construct the filter in a vector-db-specific way? @tomlynchRNA looking into this

I did some experimenting of where to pass the filters and got it working. It was a little tricky because of multiple versions of docs with similar/confusing param names, and that we don't call the method which takes the filter directly in the self query retriever, but we can fix that.

For "similarity search", this POC works:

```diff
diff --git a/agent-backend/src/tools/retrievers/similarity_search.py b/agent-backend/src/tools/retrievers/similarity_search.py
index 214231a1..3120391f 100644
--- a/agent-backend/src/tools/retrievers/similarity_search.py
+++ b/agent-backend/src/tools/retrievers/similarity_search.py
@@ -15,5 +15,13 @@ class SimilaritySearchRetriever(BaseRetriever):
     def _get_relevant_documents(self, query: str, *, run_manager: CallbackManagerForRetrieverRun):
         embedded_question = self.embedding.embed_query(query)
         if isinstance(self.vector_store, Qdrant):
-            return self.vector_store.similarity_search_with_score_by_vector(embedded_question, k=self.k)
+            from qdrant_client.http import models
+            return self.vector_store.similarity_search_with_score_by_vector(embedded_question, k=self.k, filter=models.Filter(
+                must=[
+                    models.FieldCondition(
+                        key="Retweets",
+                        match=models.MatchValue(value="11097"),
+                    )
+                ]
+            ))
         return self.vector_store.similarity_search_by_vector_with_score(embedded_question, k=self.k)
```

Just hardcodes to find a specific value, uses Must. So we could construct those filter field conditions based on some configuration in the tool that incorporates variables :sunglasses:

**Comment by iandjx** [comment]
I tried using context in this example and it worked as expected. The final output took into consideration my name, address, and email when generating the final output

```
import json
from crewai import Agent, Task, Crew
from dotenv import load_dotenv
from langchain_openai import ChatOpenAI
from pydantic import BaseModel, Field

# Define agents
user_info_agent = Agent(
    role='User Information Collector',
    goal='Collect user information including name, address, and email',
    backstory='An agent dedicated to gathering essential user information for further processing.',
)

email_generator_agent = Agent(
    role='Email Generator',
    goal='Generate an email to invite the user to talk about climate change',
    backstory="An agent dedicated to generating emails for the user's climate change inquiry.",
)

proofreader_agent = Agent(
    role='Proofreader',
    goal='Proofread and improve the email',
    backstory='An agent dedicated to proofreading and improving the email for the user.',
)

geopolitical_agent = Agent(
    role='Geopolitical Analyst',
    goal='Append geopolitical climate information based on the user’s country',
    backstory='An agent dedicated to appending geopolitical climate information based on the user’s country.',
)

# Define tasks
collect_user_info_task = Task(
    description='Ask the user for their name, address, and email',
    expected_output='User information including name, address, and email',
    agent=user_info_agent,
    human_input=True,
)

generate_email_task = Task(
    description='Generate an email to invite the user to talk about climate change. Add information about climate change.',
    expected_output='Generated email with climate change information',
    agent=email_generator_agent,
    context=[collect_user_info_task],
)

proofread_email_task = Task(
    description='Proofread the email to check for errors and improve it further',
    expected_output='Proofread and improved email',
    agent=proofreader_agent,
    context=[generate_email_task, collect_user_info_task],
)

append_geopolitical_info_task = Task(
    description='Append the geopolitical climate based on the user’s country fetched from their address',
    expected_output='Email with appended geopolitical climate information',
    agent=geopolitical_agent,
    context=[proofread_email_task, generate_email_task, collect_user_info_task],
)

# Create the crew
crew = Crew(
    agents=[user_info_agent, email_generator_agent, proofreader_agent, geopolitical_agent],
    tasks=[collect_user_info_task, generate_email_task, proofread_email_task, append_geopolitical_info_task],
    process='sequential',
)

# Kickoff the crew
crew_output = crew.kickoff()
print(crew_output)
```
